sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java

```diff
@@ -1374,11 +1419,131 @@ public void shouldNotOperateOnLocalStorageWhenPartitionNotExist() {
         assertTrue(deleteDocument.getError().getMessage().contains(failedMessage));
     }
 
+    @Test
+    public void listUserPartitionReturnsAnExceptionWhenNotSignedIn() {
+        mockSignOut();
+        PaginatedDocuments<TestDocument> documents = Data.list(TestDocument.class, USER_DOCUMENTS).get();
+        Iterator<DocumentWrapper<TestDocument>> iterator = documents.iterator();
+        assertFalse(iterator.hasNext());
+        assertNotNull(documents);
+        Page<TestDocument> currentPage = documents.getCurrentPage();
+        assertNotNull(currentPage);
+        DataException error = currentPage.getError();
+        assertNull(currentPage.getItems());
+        assertNotNull(error);
+        assertTrue(error.getMessage().contains("List operation requested on user partition, but the user is not logged in."));
+    }
+
+    private void mockSignOut() {
+        Mockito.when(mAuthTokenContext.getAccountId()).thenReturn(null);
+    }
+
+    @Test
+    public void listAnObjectWhenThereArePendingOperations() {
+        listAnObjectWhenThereArePendingOperations(Utils.getGson().toJson(new TestDocument("test")), TestDocument.class);
+    }
+
+    @Test
+    public void listPrimitiveTypeWhenThereArePendingOperations() {
+        listAnObjectWhenThereArePendingOperations("document", String.class);
+    }
+
+    private <T> void listAnObjectWhenThereArePendingOperations(String document, Class<T> documentType) {
+        Calendar expirationDate = Calendar.getInstance(TimeZone.getTimeZone("GMT"));
+        expirationDate.add(Calendar.SECOND, 1000);
+        String tokenResult = Utils.getGson().toJson(new TokenResult()
+                .setDbAccount("accountName")
+                .setDbName("dbName")
+                .setDbCollectionName("collectionName")
+                .setPartition(RESOLVED_USER_PARTITION)
+                .setExpirationDate(expirationDate.getTime())
+                .setToken("fakeToken"));
+        when(SharedPreferencesManager.getString(PREFERENCE_PARTITION_PREFIX + USER_DOCUMENTS)).thenReturn(tokenResult);
+
+        /* Return list of one item which will have a non-expired pending operation. */
+        LocalDocument localDocument = new LocalDocument(
```

Comment:
These could be private final so they could be reused between tests, but it looks like there will be another PR to improve tests?

Reply:
Yea were going to improve tests. This pr got so large were planning to do it in the other cards we have.